File: assembly/module.ts
import { register, hostCall } from "@wapc/as-guest";
import {
  Decoder,
  Writer,
  Encoder,
  Sizer,
  Codec,
  Value,
} from "@wapc/as-msgpack";

export class Host {
  binding: string;

  constructor(binding: string) {
    this.binding = binding;
  }

  update(width: u32, height: u32, limit: u32): Array<u16> {
    const inputArgs = new UpdateArgs();
    inputArgs.width = width;
    inputArgs.height = height;
    inputArgs.limit = limit;
    const payload = hostCall(
      this.binding,
      "mandelbrot",
      "update",
      inputArgs.toBuffer()
    );
    const decoder = new Decoder(payload);
    const ret = decoder.readArray(
      (decoder: Decoder): u16 => {
        return decoder.readUInt16();
      }
    );
    return ret;
  }
}

export class Handlers {
  static registerUpdate(
    handler: (width: u32, height: u32, limit: u32) => Array<u16>
  ): void {
    updateHandler = handler;
    register("update", updateWrapper);
  }
}

var updateHandler: (width: u32, height: u32, limit: u32) => Array<u16>;
function updateWrapper(payload: ArrayBuffer): ArrayBuffer {
  const decoder = new Decoder(payload);
  const inputArgs = new UpdateArgs();
  inputArgs.decode(decoder);
  const response = updateHandler(
    inputArgs.width,
    inputArgs.height,
    inputArgs.limit
  );
  const sizer = new Sizer();
  sizer.writeArray(response, (sizer: Writer, item: u16): void => {
    sizer.writeUInt16(item);
  });
  const ua = new ArrayBuffer(sizer.length);
  const encoder = new Encoder(ua);
  encoder.writeArray(response, (encoder: Writer, item: u16): void => {
    encoder.writeUInt16(item);
  });
  return ua;
}

export class UpdateArgs implements Codec {
  width: u32 = 0;
  height: u32 = 0;
  limit: u32 = 0;

  static decodeNullable(decoder: Decoder): UpdateArgs | null {
    if (decoder.isNextNil()) return null;
    return UpdateArgs.decode(decoder);
  }

  // decode
  static decode(decoder: Decoder): UpdateArgs {
    const o = new UpdateArgs();
    o.decode(decoder);
    return o;
  }

  decode(decoder: Decoder): void {
    var numFields = decoder.readMapSize();

    while (numFields > 0) {
      numFields--;
      const field = decoder.readString();

      if (field == "width") {
        this.width = decoder.readUInt32();
      } else if (field == "height") {
        this.height = decoder.readUInt32();
      } else if (field == "limit") {
        this.limit = decoder.readUInt32();
      } else {
        decoder.skip();
      }
    }
  }

  encode(encoder: Writer): void {
    encoder.writeMapSize(3);
    encoder.writeString("width");
    encoder.writeUInt32(this.width);
    encoder.writeString("height");
    encoder.writeUInt32(this.height);
    encoder.writeString("limit");
    encoder.writeUInt32(this.limit);
  }

  toBuffer(): ArrayBuffer {
    let sizer = new Sizer();
    this.encode(sizer);
    let buffer = new ArrayBuffer(sizer.length);
    let encoder = new Encoder(buffer);
    this.encode(encoder);
    return buffer;
  }
}

File: tinygo/module/module.go
package module

import (
	msgpack "github.com/wapc/tinygo-msgpack"
	wapc "github.com/wapc/wapc-guest-tinygo"
)

type Host struct {
	binding string
}

func NewHost(binding string) *Host {
	return &Host{
		binding: binding,
	}
}

func (h *Host) Update(width uint32, height uint32, limit uint32) ([]uint16, error) {
	inputArgs := UpdateArgs{
		Width:  width,
		Height: height,
		Limit:  limit,
	}
	inputBytes, err := msgpack.ToBytes(&inputArgs)
	if err != nil {
		return []uint16{}, err
	}
	payload, err := wapc.HostCall(
		h.binding,
		"mandelbrot",
		"update",
		inputBytes,
	)
	if err != nil {
		return []uint16{}, err
	}
	decoder := msgpack.NewDecoder(payload)
	listSize, err := decoder.ReadArraySize()
	if err != nil {
		return []uint16{}, err
	}
	ret := make([]uint16, 0, listSize)
	for listSize > 0 {
		listSize--
		var nonNilItem uint16
		nonNilItem, err = decoder.ReadUint16()
		if err != nil {
			return []uint16{}, err
		}
		ret = append(ret, nonNilItem)
	}
	return ret, err
}

type Handlers struct {
	Update func(width uint32, height uint32, limit uint32) ([]uint16, error)
}

func (h Handlers) Register() {
	if h.Update != nil {
		updateHandler = h.Update
		wapc.RegisterFunction("update", updateWrapper)
	}
}

var (
	updateHandler func(width uint32, height uint32, limit uint32) ([]uint16, error)
)

func updateWrapper(payload []byte) ([]byte, error) {
	decoder := msgpack.NewDecoder(payload)
	var inputArgs UpdateArgs
	inputArgs.Decode(&decoder)
	response, err := updateHandler(inputArgs.Width, inputArgs.Height, inputArgs.Limit)
	if err != nil {
		return nil, err
	}
	var sizer msgpack.Sizer
	sizer.WriteArraySize(uint32(len(response)))
	for _, v := range response {
		sizer.WriteUint16(v)
	}

	ua := make([]byte, sizer.Len())
	encoder := msgpack.NewEncoder(ua)
	encoder.WriteArraySize(uint32(len(response)))
	for _, v := range response {
		encoder.WriteUint16(v)
	}

	return ua, nil
}

type UpdateArgs struct {
	Width  uint32
	Height uint32
	Limit  uint32
}

func DecodeUpdateArgsNullable(decoder *msgpack.Decoder) (*UpdateArgs, error) {
	if isNil, err := decoder.IsNextNil(); isNil || err != nil {
		return nil, err
	}
	decoded, err := DecodeUpdateArgs(decoder)
	return &decoded, err
}

func DecodeUpdateArgs(decoder *msgpack.Decoder) (UpdateArgs, error) {
	var o UpdateArgs
	err := o.Decode(decoder)
	return o, err
}

func (o *UpdateArgs) Decode(decoder *msgpack.Decoder) error {
	numFields, err := decoder.ReadMapSize()
	if err != nil {
		return err
	}

	for numFields > 0 {
		numFields--
		field, err := decoder.ReadString()
		if err != nil {
			return err
		}
		switch field {
		case "width":
			o.Width, err = decoder.ReadUint32()
		case "height":
			o.Height, err = decoder.ReadUint32()
		case "limit":
			o.Limit, err = decoder.ReadUint32()
		default:
			err = decoder.Skip()
		}
		if err != nil {
			return err
		}
	}

	return nil
}

func (o *UpdateArgs) Encode(encoder msgpack.Writer) error {
	if o == nil {
		encoder.WriteNil()
		return nil
	}
	encoder.WriteMapSize(3)
	encoder.WriteString("width")
	encoder.WriteUint32(o.Width)
	encoder.WriteString("height")
	encoder.WriteUint32(o.Height)
	encoder.WriteString("limit")
	encoder.WriteUint32(o.Limit)

	return nil
}
